packages/0018-package-format.md

```diff
@@ -0,0 +1,163 @@
+
+Other challenges of this is about package uninstallation and how to deal with dependencies. If a NuGet package is installed in the backoffice and is dependent on 10x other NuGet packages, when we uninstall the main package we cannot also uninstall the other 10x dependencies. There may be other packages that are dependent on these packages. To keep this manageable, when working with packages in the backoffice we won't uninstall package dependencies when a package is uninstalled, like the way Visual Studio works with NuGet packages with the packages.config approach. A developer would then have the option to manually uninstall dependencies if there are no other packages dependent on it.
+
+
+#### Uniform install experience
+
+This ties into the above, but we want to ensure that the install experience is the same no matter where you install your package. Whether you drag and drop a .nupkg file into the backoffice, do it through Visual Studio or select a package in the backoffice it should feel and do the same.
+
+We want to make sure that things like the dependency management works in each case as it could otherwise open up for confusion among package consumers.
+
+#### Opt-in to changes
+
+This was mentioned briefly in the package migrations section above, but one thing we want to accomplish with this new package format is to make it more user-friendly. Currently, when you install a package there is no transparency and you have no control over what happens. Installing a package can add files, schema, content, potentially unwanted or malicious things, deletions, not handle personal data properly, etc.
+
+In an ideal world, you would install a package and get full control and insight into what it does, but that would also cause a lot of extra bloat. One thing we want to accomplish right now is to at least allow admin users to opt into when package migrations run.
+
+If the package you install or update has one or more package migrations that haven't been run you will need to authorize them with an admin account. This accomplishes two things - first is that you would have more control over things being run on the database (file changes are a bit easier to manage using version/source control such as Git).
+
+We want to show some sort of summary of what the migrations will do and present them to the user before they sign off on it.
+
+Opting in for migrations will also allow the user to plan when these things run. Can help with slow startup times due to migrations automatically running on startup.
```

Comment:
How will this work if a package uses the new package migrations to alter the database in some way and the package code is expecting these changes to be applied (e.g. a new table to exist), but the migration that would create the table isn't approved yet? This would cause the code to fail because the database is out of sync with code. This will probably just makes package developers use regular migrations instead that don't need approval... This also messes with scheduled deployments that now need a manual action before the deployment is complete. I'm not sure the advantages outweigh the disadvantages here.

Reply:
We want to avoid package developers using 'regular migrations' entirely because it's done differently by all developers which also can affect startup time depending on how people write this boilerplate code. At least here as a package developer you will never have to write the boilerplate code. If you really really really wanted your migration to run on startup, then there's nothing that can stop a package dev from doing that... but at least it will execute using Package Migrations and not a custom migration runner.

That said, we want to make all package installations consistent .... and even more important "Safe". Running things on startup is unsafe. There's nothing informing the user if anything fails, it slows down startup time, there's potential for the entirely bootup process to fail and then you're left with a YSOD anyways.

As for scheduled deployments, like i mentioned above, if there's a will there's a way and you could still run Package Migrations unassisted since there will be code to run this stuff so I suppose the power is in your hands to do that... just like we'd do on Umbraco Cloud (as mentioned) when deploying changes between environments.

Ideally if a package's files are installed and it hasn't had a required migration to run, then the package should know and handle this instead of throwing YSODs since a table might not exist. We can probably make a very simple API check for this.